### ISSUE TITLE
Aggregation without a clear final event?

### ISSUE THREAD

**Issue by phermann1988** [logstash-plugins/logstash-filter-aggregate#28]
Hi, i got a log file from a migration task and i want to aggregate logs but I have no defined final event in the loglines. is there a possible to merge the events from logdata like:

2016-05-03 15:18:55.128 [FINER] {CustomCorrelationID} Message Send to XY GroupID=T10RBE and MsgSegNumber="1" 
2016-05-03 15:18:55.487 [FINER] {CustomCorrelationID} Message Send to XY GroupID=T10RBE and MsgSegNumber="2" 

2016-05-03 15:18:56.148 [FINER] {CustomCorrelationID} Message Send to XY GroupID=X48ZUR and MsgSegNumber="1" ...

the message number debends on the groupid and varys from MsgSegNumber="1" to MsgSegNumber="645" (so no logical order like always 10 entires per groupid or sth), there is no marker for final event in the logs ;-/ the groupid entrys come one after another, so it's impossible to see before theres a line with a new GroupID and the MsgSegNumber = 1 to see theres a new migration event.

i like to aggregate like: "message" => "GroupId",  "count_msgSegNumber" => 16 (the last value of each groupid)

**Comment by alkuzad** [comment]
```
map['last_count'] ||= 0
map['last_name'] ||= ''
if map['last_name'] == event['groupID']
  map['last_count'] += 1 
else
  event['tags'] << "done"
  event['message'] = map['last_name']
  event['count_msgSegNumber'] = map['last_count']
  map['last_count'] = 0
  map['last_name'] = ''
end
```

something like that ? Then filter events on "done" tag. Task id does not matter as long as there is one stream of logs (from one file, messages are one after another)

**Comment by phermann1988** [comment]
thank you very much@alkuzad!! i will try your suggestin and give feedback

**Comment by fbaligand** [comment]
@alkuzad
it seems that your code is incomplete since `map['last_name']` is never valued with `event['groupID']`.

@phermann1988
If I understand well your need, `groupID` is your task_id (using aggregate filter vocabulary) ?
What is `CustomCorrelationID` for ?

Finally, I plan to add a new option which should exactly to your need :
`push_previous_map_as_event => true`

In case, you have no clear end event and tasks come one after other, it allows to detect that task_id has changed and automatically push previous aggregate map as a new event.

**Comment by alkuzad** [comment]
@fbaligand yep, `map['last_name'] ||= ''` should use groupID instead of empty string :)

**Comment by fbaligand** [comment]
Thanks @alkuzad for the update !

However, I doubt it works : 
I mean : when you add "done" tag, you add it on next groupid and not previous task id,,which is done.

**Comment by alkuzad** [comment]
@fbaligand this is why event is altered manually, user can filter on done tag to get rid of current groupID and send only to output also on done tag. This is not trivial but I have done something similar to connect several messages together (with additional external queue) and it works.

**Comment by fbaligand** [comment]
I mean : when you enter in else part, event groupId is different from map groupId, so current event is on next group, so you add tag "done" on next group. 

That said, you use a trick : you transform current event in previous group using map :)

**Comment by fbaligand** [comment]
This issue is now fixed with new release 2.2.0.
It adds a new option : push_previous_map_as_event

You can see a good example (Example # 3) in [README.md](https://github.com/logstash-plugins/logstash-filter-aggregate#example-3) page.

**Comment by yaslamenace** [comment]
+1, I am very interested in this feature as explained by alkuzad !
